### DISCUSSION TITLE
Meta language supports APE

### DISCUSSION THREAD

**Discussion by Kaj-de-Vos** [jart/cosmopolitan#718] [Show and tell]
When a customer of mine pointed me to APE, my jaw spent the next few days on the floor. I subsequently used it to compile my new programming language to PC targets, enabling me to cover all APE targets at once:

[https://language.metaproject.frl](https://language.metaproject.frl) 

I have to admit that I ran a little experiment by not telling about APE at first. Hardly anyone remarked about why there were no separate binaries for different platforms. Almost everyone seems to give no thought to it, considering it natural, even in programmer circles.

Meta is a descendant of the REBOL language, which was itself inspired by Logo, Lisp and Forth. It's a young language, that I design to be adaptable, so it doesn't have the porting problems to APE of existing languages. It is developing together with APE and Cosmopolitan to implement features similar to REBOL and beyond.

Meta has multiple back-ends, but the default is to transpile to C, to have the widest possible reach. This makes it a good match to target APE and Cosmopolitan.

Instead of GCC, Meta uses CLang with APE, because GCC computes function arguments in reverse order. Using CLang produces a natural left-to-right evaluation order, compatible with REBOL.

**Comment by jart** [comment]
> I have to admit that I ran a little experiment by not telling about APE at first. Hardly anyone remarked about why there were no separate binaries for different platforms. Almost everyone seems to give no thought to it, considering it natural, even in programmer circles.

That's remarkable. Thank you for sharing this.

> Instead of GCC, Meta uses CLang with APE, because GCC computes function arguments in reverse order. Using CLang produces a natural left-to-right evaluation order, compatible with REBOL.

One thing you can do, if Clang isn't installed on the user's PC, is you can embed [cosmopolitan/third_party/chibicc/](https://github.com/jart/cosmopolitan/tree/master/third_party/chibicc) which supports C11 and a great number of GCC extensions, however the code it compiles goes 2x slower since it currently doesn't have an optimizer. The missing piece though is we need a linker. So far we've built a compiler and assembler, but we should ideally have some kind of `ld` replacement that can emit the APE format.

**Reply by Kaj-de-Vos** [reply]
Hi Justine. Thank you very much for several awesome projects!

I've been interested in the C compiler you support. Do you happen to know the computation order of function arguments it compiles?

It's not urgent for Meta, though. Because of such intricacies of multiple used toolchains, the Meta compiler is offered as a web service that we maintain. However, a suitable simple C toolchain would be an option to eventually offer a locally installable version.

**Reply by jart** [reply]
I'm reasonably certain it's implemented as left-to-right https://github.com/jart/cosmopolitan/blob/be3e109309d7adc0824b8c22cf4fca3aa6433baa/third_party/chibicc/codegen.c#L711 noting that this isn't documented or guaranteed to not change if we add an optimizer, since argument evaluation order is undefined behavior per the C standard.

**Reply by jart** [reply]
Actually I take that back. Looking at the code it's not immediately obvious. But that's where you'd look to find out.

**Reply by Kaj-de-Vos** [reply]
Thanks. Yes, I have to curate C compilers for suitability. It's good news that ChibiCC is a candidate. Changing computation order without the compiler supporting it would be a massive effort.
